### PR TITLE
fix: match with same documents

### PR DESCRIPTION
**Pull request by bwanglzu**
for #3034 , I would not say it's a bug, it's a side effect. Calling match on same document does not really make sense. But here is a workaround. not a perfect solution, but works. Feel free to close this if there is a better solution

before: 
1D590F4A-B5CB-47C4-81F5-A30E53B4354B

after:

3FE4B9E8-90EF-4729-9F1B-687952086F2E

### CODE REVIEW

**Review by tadejsv — DISMISSED** [review]
Looks good to me

File: jina/types/arrays/neural_ops.py [hunk 1]
```diff
@@ -75,10 +75,16 @@ def match(
             if isinstance(normalization, (tuple, list)):
                 dist = minmax_normalize(dist, normalization)
 
+        source_docarray_ids = [d.id for d in self]
```
Comment (hanxiao) [inline jina/types/arrays/neural_ops.py:78]: use set

File: jina/types/arrays/neural_ops.py [hunk 2]
```diff
@@ -75,10 +75,16 @@ def match(
...
         m_name = metric_name or (metric.__name__ if callable(metric) else metric)
         for _q, _ids, _dists in zip(self, idx, dist):
             _q.matches.clear()
             for _id, _dist in zip(_ids, _dists):
                 d = Document(darray[int(_id)], copy=True)
+                # Note, when match self with other, or both of them share the same Document
+                # we might have recursive matches .
+                # checkout https://github.com/jina-ai/jina/issues/3034
+                if d.id in source_docarray_ids:
```
Comment (hanxiao) [inline jina/types/arrays/neural_ops.py:87]: list is low-efficient

**Review by hanxiao — CHANGES_REQUESTED** [review]
do not merge

**Pull request merged**